### PR TITLE
feat: [PL-39761]: added new EULA svc to sign different user agreements

### PR DESCRIPTION
**Pull request by bhavya181**
## Describe your changes
 added new EULA svc to sign different user agreements

Tested - 
1. checked validate EULA is already signed or not API 
2. signed eula for AIDA, audit event is also sent

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51136)

### CODE REVIEW

File: 120-ng-manager/modules/eula/src/main/java/io/harness/eula/service/impl/EulaServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  public boolean sign(EulaDTO eulaDTO) {
+    Eula newEula = eulaMapper.toEntity(eulaDTO);
+    Optional<Eula> existingEula = get(eulaDTO.getAccountIdentifier());
+    if (existingEula.isPresent()) {
+      Set<AgreementType> signedAgreements = existingEula.get().getSignedAgreements();
+      signedAgreements.add(eulaDTO.getAgreement());
+      newEula.setSignedAgreements(signedAgreements);
+    }
+    return Failsafe.with(DEFAULT_RETRY_POLICY).get(() -> transactionTemplate.execute(status -> {
+      Eula eula = eulaRepository.upsert(newEula);
```
Comment (gn-harness) [inline 120-ng-manager/modules/eula/src/main/java/io/harness/eula/service/impl/EulaServiceImpl.java:50]: nit: Add an info message with details on the eula sign event

File: 120-ng-manager/src/main/java/io/harness/ng/EulaModule.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+public class EulaModule extends AbstractModule {
+  NextGenConfiguration appConfig;
+
+  public EulaModule(NextGenConfiguration appConfig) {
+    this.appConfig = appConfig;
+  }
+
+  @Override
+  protected void configure() {
+    bind(NextGenConfiguration.class).toInstance(appConfig);
```
Comment (gn-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/EulaModule.java:24]: Why is this required again? Some other module should already have been binding this, right?

**Review by gn-harness — APPROVED** [review]

File: 120-ng-manager/modules/eula/src/main/java/io/harness/eula/service/impl/EulaServiceImpl.java (hunk 1 above)
Comment (gn-harness) [reply 120-ng-manager/modules/eula/src/main/java/io/harness/eula/service/impl/EulaServiceImpl.java:50]: Also add a few simple tests

**Review by gn-harness — DISMISSED** [review]

**Review by bhavya181 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51136)* status: 43 file(s) to review. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`120-ng-manager/modules/eula/src/main/java/io/harness/eula/events/EulaSignEvent.java` line 48 at r2](https://reviewable.io/reviews/harness/harness-core/51136#gh-1281233566:-Napzdsp7O7XvWc4A8sy:brs1iga) ([raw file](https://github.com/harness/harness-core/blob/a71a08cb237db986a7144f232e9d2a64427899ac/120-ng-manager/modules/eula/src/main/java/io/harness/eula/events/EulaSignEvent.java#L48)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

What do these labels represent?
</blockquote></details>

labels are basically resource parameters that needs to be passed to audit service, here Eula only we can send name of resource same as agreement type.
___
*[`120-ng-manager/modules/eula/src/main/java/io/harness/eula/service/EulaService.java` line 21 at r2](https://reviewable.io/reviews/harness/harness-core/51136#gh-1281255968:-NapzuI--wSUO3u7I62G:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/a71a08cb237db986a7144f232e9d2a64427899ac/120-ng-manager/modules/eula/src/main/java/io/harness/eula/service/EulaService.java#L21)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

nit: Add some javadocs to the interface. You can add some context around what is the service for and how it can be extended etc
</blockquote></details>

Done
___
*[`120-ng-manager/modules/eula/src/main/java/io/harness/eula/service/impl/EulaServiceImpl.java` line 50 at r2](https://reviewable.io/reviews/harness/harness-core/51136#gh-1281256982:-NapzvsM30Faxe44Suyg:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/a71a08cb237db986a7144f232e9d2a64427899ac/120-ng-manager/modules/eula/src/main/java/io/harness/eula/service/impl/EulaServiceImpl.java#L50)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

Also add a few simple tests
</blockquote></details>

Done
___
*[`120-ng-manager/src/main/java/io/harness/ng/EulaModule.java` line 24 at r2](https://reviewable.io/reviews/harness/harness-core/51136#gh-1281264432:-Napzxy9Ey4kVOlQSST7:b-pykaxi) ([raw file](https://github.com/harness/harness-core/blob/a71a08cb237db986a7144f232e9d2a64427899ac/120-ng-manager/src/main/java/io/harness/ng/EulaModule.java#L24)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

Why is this required again? Some other module should already have been binding this, right?
</blockquote></details>

Yes, NextGenModule is there, but keeping all configs is making that file large and difficult that's why creating new modules, then registering this module in NextGenModule.

**Review by gn-harness — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51136)* status: 43 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`120-ng-manager/modules/eula/src/test/java/io/harness/eula/service/impl/EulaServiceImplTest.java` line 57 at r5](https://reviewable.io/reviews/harness/harness-core/51136#-NarAer_8GV6NdDsG_ow:-NarAer_8GV6NdDsG_ox:bwl2lod) ([raw file](https://github.com/harness/harness-core/blob/35ab0523162c90478cb41709b61cccb60523cf6f/120-ng-manager/modules/eula/src/test/java/io/harness/eula/service/impl/EulaServiceImplTest.java#L57)):*
> ```Java
>   @Owner(developers = BHAVYA)
>   @Category(UnitTests.class)
>   public void testSignEula_whenAgreementIsAlreadySigned() {
> ```

nit: name your tests like "testName_setup_resultingState"

something on the lines of: signEula_agreementAlreadySigned_NoNewEulaSignEvent
___
*[`120-ng-manager/src/main/java/io/harness/ng/EulaModule.java` line 24 at r2](https://reviewable.io/reviews/harness/harness-core/51136#gh-1281264432:-NarB_px5z-Q4JglYpre:b-ebtgno) ([raw file](https://github.com/harness/harness-core/blob/a71a08cb237db986a7144f232e9d2a64427899ac/120-ng-manager/src/main/java/io/harness/ng/EulaModule.java#L24)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

Yes, NextGenModule is there, but keeping all configs is making that file large and difficult that's why creating new modules, then registering this module in NextGenModule.
</blockquote></details>

Creating new modules is fine, actually preferred. My question was on passing in the appConfig and binding that to NextGenConfiguration.class

**Comment by bhavya181** [comment]
trigger uts

**Comment by bhavya181** [comment]
trigger uts

**Comment by bhavya181** [comment]
trigger ut2

**Comment by bhavya181** [comment]
trigger uts

**Comment by bhavya181** [comment]
trigger ut5

**Comment by bhavya181** [comment]
trigger ut

**Comment by bhavya181** [comment]
trigger uts

**Comment by bhavya181** [comment]
trigger ut7

**Comment by bhavya181** [comment]
trigger uts

**Comment by bhavya181** [comment]
trigger ut2, ut7, ut9

**Comment by bhavya181** [comment]
trigger ut7, ut9

**Comment by bhavya181** [comment]
trigger uts

**Comment by bhavya181** [comment]
trigger ut1, ut7
